Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+StateMachine.swift

```diff
@@ -0,0 +1,287 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+extension HTTPConnectionPool {
+    struct StateMachine {
+        struct Action {
+            let request: RequestAction
+            let connection: ConnectionAction
+
+            init(request: RequestAction, connection: ConnectionAction) {
+                self.request = request
+                self.connection = connection
+            }
+
+            static let none: Action = Action(request: .none, connection: .none)
+        }
+
+        enum ConnectionAction {
+            enum IsShutdown: Equatable {
+                case yes(unclean: Bool)
+                case no
+            }
+
+            case createConnection(Connection.ID, on: EventLoop)
+            case scheduleBackoffTimer(Connection.ID, backoff: TimeAmount, on: EventLoop)
+
+            case scheduleTimeoutTimer(Connection.ID, on: EventLoop)
+            case cancelTimeoutTimer(Connection.ID)
+
+            case closeConnection(Connection, isShutdown: IsShutdown)
+            case cleanupConnections(CleanupContext, isShutdown: IsShutdown)
+
+            case none
+        }
+
+        enum RequestAction {
+            case executeRequest(Request, Connection, cancelTimeout: Request.ID?)
+            case executeRequests([(Request, cancelTimeout: Request.ID?)], Connection)
+
+            case failRequest(Request, Error, cancelTimeout: Request.ID?)
+            case failRequests([(Request, cancelTimeout: Request.ID?)], Error)
+
+            case scheduleRequestTimeout(NIODeadline?, for: Request.ID, on: EventLoop)
+            case cancelRequestTimeout(Request.ID)
+
+            case none
+        }
+
+        enum HTTPTypeStateMachine {
+            case http1(HTTP1StateMachine)
+
+            case _modifying
```

Comment:
We almost certainly shouldn't need this case. I know I haven't provided this feedback much prior to here, so don't feel compelled to remove it now, but we should probably go through these state machines and rip out the `_modifying` states wherever we have then. Recent Swift versions do not CoW here.